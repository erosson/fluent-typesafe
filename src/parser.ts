import * as Syntax from '@fluent/syntax'
import { promises as fs } from 'fs'
import { uniqBy } from 'lodash'

export type Resource = {
    path: string,
    raw: string,
    resource: Syntax.Resource,
    messages: Message[],
}
export type Message = {
    id: string,
    placeholders: Var[],
    attributes: Set<string>,
    raw: string,
}
export type Var = {
    name: string,
    type: VarType,
}
export enum VarType {
    STRING = 'string',
    NUMBER = 'number',
    DATETIME = 'datetime',
}
function parseInlineExpression(expr: Syntax.InlineExpression): Var | null {
    switch (expr.type) {
        case 'VariableReference':
            return { name: expr.id.name, type: VarType.STRING }
        case 'FunctionReference':
            const first = expr.arguments.positional[0]
            switch (first?.type) {
                case 'VariableReference':
                    const name = first.id.name
                    switch (expr.id.name) {
                        case 'NUMBER':
                            return { name, type: VarType.NUMBER }
                        case 'DATETIME':
                            return { name, type: VarType.DATETIME }
                        default: return null
                    }
                default: return null
            }
        default: return null
    }
}
function parseExpression(expr: Syntax.Expression): Var[] {
    switch (expr.type) {
        case 'SelectExpression':
            return [parseInlineExpression(expr.selector), ...expr.variants.map(variant => parsePattern(variant.value)).flat()].filter((v): v is Var => !!v)
        default: return [parseInlineExpression(expr)].filter((v): v is Var => !!v)
    }
}
function parsePattern(pattern: Syntax.Pattern): Var[] {
    const placeables = pattern.elements.filter((el): el is Syntax.Placeable => el.type === 'Placeable')
    return placeables.map(el => parseExpression(el.expression)).flat().filter((exp): exp is Var => !!exp)
}
function parseResource_(path: string, raw: string): Resource {
    const resource = Syntax.parse(raw, {})
    const msgs: Syntax.Message[] = resource.body.filter<Syntax.Message>((entry): entry is Syntax.Message => entry.type === 'Message')
    const messages: Message[] = msgs.map(message => {
        const id = message.id.name
        const attributes = new Set(message.attributes.map(attr => attr.id.name))
        const patterns: Syntax.Pattern[] = [message.value, ...message.attributes.map(a => a.value)].filter((p): p is Syntax.Pattern => !!p)
        // TODO prefer numbers and datetimes over strings
        const placeholders: Var[] = uniqBy(patterns.map(parsePattern).flat(), 'name')
        const rawMsg = raw.slice(message.span.start, message.span.end)
        return { id, placeholders, attributes, raw: rawMsg }
    })
    return { path, raw, resource, messages }
}
export async function parseResource(path: string): Promise<Resource> {
    const raw = await fs.readFile(path, 'utf8')
    return parseResource_(path, raw)
}
