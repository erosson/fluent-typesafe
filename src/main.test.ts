import * as Parser from "./parser"
import ElmFormat from "./codegen/elm"
import ReactFormat from "./codegen/react"
import ReactDomFormat from "./codegen/react-dom"

test("simple", async () => {
    const path = "app.ftl"
    const ftl = `\
hello = world
`
    const parsed: Parser.Resource = Parser.parseResource_(path, ftl)
    expect(parsed).toMatchSnapshot()
    for (let { label, format } of [{ label: 'elm', format: ElmFormat }, { label: 'react', format: ReactFormat }, { label: 'react-dom', format: ReactDomFormat }]) {
        const writer = jest.fn()
        await format([[path, Promise.resolve(parsed)]], writer)
        expect(writer.mock.calls.length).toBe(2)
        // consistent windows/linux path behavior
        const calls = writer.mock.calls.map(([path, body]) => [path.replace('\\', '/'), body])
        expect({ label, calls }).toMatchSnapshot()
    }
})