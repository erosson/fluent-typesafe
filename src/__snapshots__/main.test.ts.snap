// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple 1`] = `
Object {
  "messages": Array [
    Object {
      "attributes": Set {},
      "id": "hello",
      "placeholders": Array [],
      "raw": "hello = world",
    },
  ],
  "path": "app.ftl",
  "raw": "hello = world
",
  "resource": Resource {
    "body": Array [
      Message {
        "attributes": Array [],
        "comment": null,
        "id": Identifier {
          "name": "hello",
          "span": Span {
            "end": 5,
            "start": 0,
            "type": "Span",
          },
          "type": "Identifier",
        },
        "span": Span {
          "end": 13,
          "start": 0,
          "type": "Span",
        },
        "type": "Message",
        "value": Pattern {
          "elements": Array [
            TextElement {
              "span": Span {
                "end": 13,
                "start": 8,
                "type": "Span",
              },
              "type": "TextElement",
              "value": "world",
            },
          ],
          "span": Span {
            "end": 13,
            "start": 8,
            "type": "Span",
          },
          "type": "Pattern",
        },
      },
    ],
    "span": Span {
      "end": 14,
      "start": 0,
      "type": "Span",
    },
    "type": "Resource",
  },
}
`;

exports[`simple 2`] = `
Object {
  "calls": Array [
    Array [
      "Localization/App.elm",
      "module Localization.App exposing (hello, name_)

{-| File auto-generated by \`@erosson/fluent-typesafe\`. Do not edit!
-}

import Html as H
import Html.Attributes as A


name_ : String -> H.Attribute msg
name_ =
    A.attribute \\"data-l10n-name\\"


{-| Fluent message id \`hello\`, in file \`app.ftl\`

    hello =
        world

-}
hello : H.Attribute msg
hello =
    A.attribute \\"data-l10n-id\\" \\"hello\\"
",
    ],
    Array [
      "Localization.elm",
      "module Localization exposing (app)

{-| File auto-generated by \`@erosson/fluent-typesafe\`. Do not edit!
-}

import Localization.App


app =
    { hello = Localization.App.hello
    }
",
    ],
  ],
  "label": "elm",
}
`;

exports[`simple 3`] = `
Object {
  "calls": Array [
    Array [
      "localization/app.tsx",
      "/** File auto-generated by \`@erosson/fluent-typesafe\`. Do not edit! */

// eslint-disable-next-line @typescript-eslint/no-unused-vars
import __React from \\"react\\";
import * as __L from \\"@fluent/react\\";

// https://github.com/projectfluent/fluent.js/blob/master/fluent-react/src/localized.ts
type __props = Omit<__L.LocalizedProps, \\"id\\" | \\"vars\\" | \\"attrs\\">;

/** Fluent message id \`hello\`, in file \`app.ftl\`

\`\`\`
hello = world
\`\`\`

*/

export function Hello(props: __props): JSX.Element {
  return <__L.Localized id=\\"hello\\" {...props} />;
}
",
    ],
    Array [
      "localization.ts",
      "/** File auto-generated by \`@erosson/fluent-typesafe\`. Do not edit! */

import * as localizationApp from \\"./localization/app\\";

export const app = localizationApp;
",
    ],
  ],
  "label": "react",
}
`;

exports[`simple 4`] = `
Object {
  "calls": Array [
    Array [
      "localization/app.ts",
      "/** File auto-generated by \`@erosson/fluent-typesafe\`. Do not edit! */

export function name_(name: string): { \\"data-l10n-name\\": string } {
  return { \\"data-l10n-name\\": name };
}

/** Fluent message id \`hello\`, in file \`app.ftl\`

\`\`\`
hello = world
\`\`\`

*/

export const hello: { \\"data-l10n-id\\": string } = { \\"data-l10n-id\\": \\"hello\\" };
",
    ],
    Array [
      "localization.ts",
      "/** File auto-generated by \`@erosson/fluent-typesafe\`. Do not edit! */

import * as localizationApp from \\"./localization/app\\";

export const app = localizationApp;
",
    ],
  ],
  "label": "react-dom",
}
`;
